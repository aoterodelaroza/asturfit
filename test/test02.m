#! /usr/bin/octave -q
% Copyright (C) 2010 Victor Lua~na and Alberto Otero-de-la-Roza
%
% This octave routine is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or (at
% your option) any later version. See <http://www.gnu.org/licenses/>.
%
% The routine distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
% FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
% more details.

function status = test02 (filein, rootname)
%
% TEST02: Standard fit to a set of data files.
%
% Run as:   test02.m file1.dat file2.dat ...

global nelectrons
hybohr3togpa = 2*14710.50498740275538944426;
status = 0;

addpath("../src/");

printf("+++++++++++++++++++++++++++++++\n");
printf("test02: filein and rootname --> %s, %s\n", filein, rootname);
printf("+++++++++++++++++++++++++++++++\n");

[v,e] = readEVdata(filein,1);
vref = median(v);

# Strain fitting. Decide up to wich degree do the average.
n = 4;
newerr = ones(4,1)*1e10;
rk = [1 3 4 5];
strain = "eulerian";
printf("Fit to polynomials of %s strain\n", strain);
printf("Select the best number of polynomials to average:\n");
printf("Checking err(X) = (std(X)/mean(X))^2 for:\n");
printf(" --n-- --volume-- ----B----- ----B1p--- ----B2p---\n");
do
   olderr = newerr;
   [cf,sf] = avgstrainfit(v,e,vref,n,:,strain,0);
   newerr = (sf.eqstd(rk)./sf.eqmean(rk)).^2;
   printf("%6d", n);
   printf("  %.3e", newerr);
   printf("\n");
   n++;
until (sum(newerr > 1.5*olderr)>1 | n > 18)
[cf,sf] = avgstrainfit(v,e,vref,--n,:,strain,1);

# Figures will not appear on screen:
set(gcf(),"visible","off");

# Figures: E(V), p(V), B(p), B(V), B1(p), B*B2(p)
fileplot = sprintf("%sev.eps", rootname);
plot(v,e,"ob", v,sf.Efit,"-r");
xlabel("V (bohr^3)"); ylabel("E (hartree)"); grid("on");
print(fileplot, "-FHelvetica:22", "-depsc");
printf("E(V) plot created in: %s\n", fileplot);

fileplot = sprintf("%spv.eps", rootname);
f = hybohr3togpa;
errorbar(v, sf.pmean*f, sf.pstd*f, "~");
xlabel("V (bohr^3)"); ylabel("p (GPa)"); grid("on");
print(fileplot, "-FHelvetica:22", "-depsc");
printf("p(V) plot created in: %s\n", fileplot);

fileplot = sprintf("%sbp.eps", rootname);
f = hybohr3togpa;
errorbar(sf.pmean*f, sf.Bmean*f, sf.pstd*f, sf.Bstd*f, "~>");
xlabel("p (GPa)"); ylabel("B (GPa)"); grid("on");
print(fileplot, "-FHelvetica:22", "-depsc");
printf("B(p) plot created in: %s\n", fileplot);

fileplot = sprintf("%sbv.eps", rootname);
plot(v,sf.Bmean*f,"-ob");
xlabel("V (bohr^3)"); ylabel("B (GPa)"); grid("on");
print(fileplot, "-FHelvetica:22", "-depsc");
printf("B(V) plot created in: %s\n", fileplot);

fileplot = sprintf("%sb1p.eps", rootname);
errorbar(sf.pmean*f, sf.B1pmean, sf.pstd*f, sf.B1pstd, "~>");
xlabel("p (GPa)"); ylabel("B^'"); grid("on");
print(fileplot, "-FHelvetica:22", "-depsc");
printf("B1(p) plot created in: %s\n", fileplot);

fileplot = sprintf("%sbb2p.eps", rootname);
errorbar(sf.pmean*f, sf.Bmean.*sf.B1pmean, sf.pstd*f, sf.Bstd.*sf.B1pstd, "~>");
xlabel("p (GPa)"); ylabel("B B^{''}"); grid("on");
print(fileplot, "-FHelvetica:22", "-depsc");
printf("B*B2p(p) plot created in: %s\n", fileplot);

# Reactivate the viewing on screen:
set(gcf(),"visible","on");

endfunction

## Use it as a script
if (!exist("argn"))
   if (nargin > 0)
      args = argv();
      for i = 1 : nargin
         test02(args{i}, sprintf("test02-c%02d-",i));
      endfor
   else
      printf("Use as: test02.m file(s)\n");
   endif
endif
