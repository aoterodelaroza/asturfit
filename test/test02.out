warning: mark_as_command is obsolete and will be removed from a future version of Octave
+++++++++++++++++++++++++++++++
test02: filein and rootname --> w2k-lda-na.dat, test02-c01-
+++++++++++++++++++++++++++++++
readEVdata: Reading input file -> w2k-lda-na.dat
Input volume unit: bohr^3
Input energy unit: ry
* Energy and volume ranges 
  V-range (bohr^3) : [159.9086000000 , 319.8173000000] dV = 159.9087000000
  E-range (hartree) : [-161.7088550000 , -161.7039449000] dE = 0.0049101000
  Number of data points: 61
  Number of electrons: 11

Fit to polynomials of eulerian strain
Select the best number of polynomials to average:
Checking err(X) = (std(X)/mean(X))^2 for:
 --n-- --volume-- ----B----- ----B1p--- ----B2p---
     4  7.764e-10  2.685e-06  1.460e-06  2.640e-03
     5  4.736e-10  2.009e-06  9.780e-07  2.109e-03
     6  3.527e-10  1.514e-06  8.320e-07  1.547e-03
     7  4.655e-10  1.229e-06  3.118e-06  1.501e-03
     8  4.490e-10  1.086e-06  3.254e-06  1.874e-03
     9  4.063e-10  9.599e-07  2.917e-06  1.794e-03
    10  4.599e-10  8.313e-07  5.908e-06  3.337e-03


avgpolyfit: AVERAGE OF eulerian STRAIN POLYNOMIALS
Volume reference (V0): 239.863000
Range of degrees: 2--10
Number of polynomials: 9

Properties at the minimum of each polynomial:
--i- npol data npar --SSerr-- ---w---- ----Vmin--- ----Emin--- ----Bmin--- ---B1min--- ---B2min---
   1    3   61    4  8.65e-12 0.138405  224.457562 -161.708856    9.163736    3.730992 -0.361829533 0.304202700
   2    4   61    5  7.98e-12 0.136732  224.451813 -161.708856    9.163018    3.736722 -0.359771200 0.293634210
   3    5   61    6  7.90e-12 0.134412  224.450837 -161.708856    9.164838    3.738213 -0.365880463 0.293260394
   4    6   61    7  7.40e-12 0.133058  224.444541 -161.708856    9.162432    3.750303 -0.353090715 0.235454328
   5    7   61    8  7.37e-12 0.130907  224.445704 -161.708856    9.160572    3.747259 -0.341629796 0.245756656
   6    8   61    9  7.36e-12 0.128736  224.446565 -161.708856    9.161242    3.744563 -0.347399272 0.269324175
   7    9   61   10  7.16e-12 0.127162  224.442281 -161.708856    9.167272    3.761747 -0.405833579 0.143169052
   8    2   61    3  1.10e-10 0.070587  224.444566 -161.708857    9.194978    3.740368 -0.402030944 0.322548165
   9    1   61    2  2.44e-08 0.000000  223.719254 -161.708856    9.221046    4.000000 -0.421740528 0.365893864

Average properties (weighted polynomials):
------ ---volume-- ---energy-- --B-(GPa)-- ----B1p---- B2p-(1/GPa) B3p--(1/GPa^2)
-mean-  224.448311 -161.708856    9.165532    3.743765   -0.364861    0.260815760
stdvev    0.004813    0.000002    0.008357    0.009100    0.021078    0.051989496
E(V) plot created in: test02-c01-ev.eps
p(V) plot created in: test02-c01-pv.eps
B(p) plot created in: test02-c01-bp.eps
B(V) plot created in: test02-c01-bv.eps
B1(p) plot created in: test02-c01-b1p.eps
B*B2p(p) plot created in: test02-c01-bb2p.eps
